Sauce Visual API lifecycle (#2867)
saucelabs/sauce-docs: docs/visual-testing/workflows/api-lifecycle.md

## What You'll Need

- A Sauce Labs account ([Log in](https://accounts.saucelabs.com/am/XUI/#login/) or sign up for a [free trial license](https://saucelabs.com/sign-up))
- Familiarity with GraphQL queries and mutations
- Tools like Postman or cURL for testing API calls

## Lifecycle Steps

### 1. Create a Build

To start, you need to create a build. This initializes the process and prepares the environment for subsequent steps.

**GraphQL Mutation:**

```graphql
mutation {
  createBuild(input: { name: "Your Build Name" }) {
    id
    name
    status
    url
  }
}
```

**Expected Response:**

```json
{
  "data": {
    "createBuild": {
      "id": "build-id-here",
      "name": "Your Build Name",
      "status": "RUNNING",
      "url": "https://app.saucelabs.com/visual/builds/build-id-here"
    }
  }
}
```

- `status`: As a newly created build without any snapshots, this will be `RUNNING`.
- `url`: The URL that can be used any time to check the build on Sauce Labs.

### 2. Upload Image

Next, upload an image to the build. This is a two step process. 

> First, get a placeholder URL for image upload using `createSnapshotUpload`.

Reviewer: ```suggestion
First, obtain a signed URL for uploading your image by using the `createSnapshotUpload` mutation.
```

**GraphQL Mutation:**

```graphql
mutation {
  createSnapshotUpload(input: {buildId: "build-id-here"}) {
    id

>     uploadUrl

Reviewer: Shouldn't we mention `domUploadUrl` too ?

  }
}
```

- `buildId`: The ID of the build created in the previous step.

**Expected Response:**

```json
{
  "data": {
    "createSnapshotUpload": {
      "id": "upload-id-here",
      "uploadUrl": "upload-url-here"
    }
  }
}
```

- `id`: Upload ID to use in the subsequent steps.
- `uploadUrl`: The URL to upload the image in the next step.

> Next, send a `PUT` request to `uploadUrl` with image file in the body of the request. Only **PNG** files are supported.

Reviewer: Should we mention the support for `Content-MD5` header ?

It's useful when it comes to ensuring the image has been correctly uploaded

**cURL Representation:**

```sh
curl --request PUT \
  --url 'upload-url-here' \
  --header 'Content-Type: image/png' \

>   --data 'png-file-content'

Reviewer: Alternative representation:

```suggestion
  --data '@my-screenshot.png'
```

(Starting with `@` will read the file).

```

### 3. Create Snapshot

> After uploading an image, create a snapshot to capture the current state of the build.

Reviewer: I find that part not super clear. What actually means "to capture the current state of the build" ?

Suggestion (may need improvements):
```suggestion
After uploading your image, add your snapshot and its metadata to your build.
```

**GraphQL Mutation:**

```graphql
mutation {
  createSnapshot(

>     input: {buildUuid: "build-id-here", uploadId: "upload-id-here", name: "Your snapshot name"}

Reviewer: We should mention at least `branch`, `project` fields. That's dimensions that we expect users to use.

We should also strongly advise to fill in `OS` / `Browser`.

Author: `branch` and `project` are build specific properties, not snapshot. I'll mention them there.

  ) {
    id
    uploadId
  }
}
```

- `uploadId`: Upload ID acquired with `createSnapshotUpload` in the previous step.
- `buildUuid`: Build ID that was used in previous steps.

**Expected Response:**

```json
{
  "data": {
    "createSnapshot": {
      "id": "snapshot-id-here",
      "uploadId": "upload-id-here"
    }
  }
}
```

- `id`: The ID of the snapshot that has been created.

> ### 4. Finish Build

Reviewer 2: - can we add information we can close the build automatically after a few hours?
- can we point out this is necessary step and the build cannot be re-used after closing it?

> Finally, finish the build process to mark it as complete.

Reviewer: Probably better this way:
```suggestion
Finally, finish the build to mark it as complete.
```
